fenicsxconcrete/experimental_setup/concrete_cylinder.py

```diff
@@ -0,0 +1,251 @@
+        bc_generator = BoundaryConditions(self.mesh, V)
+
+        if self.p['bc_setting'] == "fixed":
+            if self.p['dim'] == 2:
+                bc_generator.add_dirichlet_bc(self.top_displacement, self.boundary_top(), 1, "geometrical", 1)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), self.boundary_top(), 0, "geometrical", 0)
+                bc_generator.add_dirichlet_bc(df.fem.Constant(domain=self.mesh, c=(0, 0)), self.boundary_bottom(), "geometrical")
+            elif self.p['dim']== 3:
+
+                bc_generator.add_dirichlet_bc(self.top_displacement, self.boundary_top(), 2, "geometrical", 2)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), self.boundary_top(), 0, "geometrical", 0)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), self.boundary_top(), 1, "geometrical", 1)
+                bc_generator.add_dirichlet_bc(df.fem.Constant(domain=self.mesh, c=(0, 0, 0)), self.boundary_bottom(), "geometrical")
+
+        elif self.p['bc_setting'] == "free":
+            if self.p['dim'] == 2:
+                bc_generator.add_dirichlet_bc(self.top_displacement, self.boundary_top(), 1, "geometrical", 1)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), self.boundary_bottom(), 1, "geometrical", 1)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), point_at((0, 0)), 0, "geometrical", 0)
+
+            elif self.p['dim'] == 3:
+                # getting nodes at the bottom of the mesh to apply correct boundary condition to arbitrary cylinder mesh
+                mesh_points = self.mesh.geometry.x  # list of all nodal coordinates
+                bottom_points = mesh_points[(mesh_points[:, 2] == 0.0)]  # copying the bottom nodes, z coord = 0.0
+
+                # sorting by x coordinate
+                x_min_boundary_point = bottom_points[bottom_points[:, 0].argsort(kind="mergesort")][0]
+                x_max_boundary_point = bottom_points[bottom_points[:, 0].argsort(kind="mergesort")][-1]
+                # sorting by y coordinate
+                y_boundary_point = bottom_points[bottom_points[:, 1].argsort(kind="mergesort")][0]
+            
+                bc_generator.add_dirichlet_bc(self.top_displacement, self.boundary_top(), 2, "geometrical", 2)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), self.boundary_bottom(), 2, "geometrical", 2)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), point_at(x_min_boundary_point), 1, "geometrical", 1)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), point_at(x_max_boundary_point), 1, "geometrical", 1)
+                bc_generator.add_dirichlet_bc(np.float64(0.0), point_at(y_boundary_point), 0, "geometrical", 0)
+            
+            else:
+                raise Exception(f"dim setting: {self.p['dim']}, not implemented for cylinder bc setup: free")
+        else:
+            raise Exception(f"Wrong boundary setting: {self.p['bc_setting']}, for cylinder setup")
```

Comment:
these exceptions are not tested. please add them to the tests.
an example of how to test for Errors is found in: https://github.com/BAMresearch/FenicsXConcrete/blob/47df6466911fc6f7b64ba080137ef518f6ceadf6/tests/experimental_setup/test_experimental_setups.py#L37
Changing the Exceptopn to a ValueError might be more informative.

Reply:
Added tests